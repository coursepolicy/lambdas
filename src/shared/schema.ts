import { z } from 'zod';
import { PolicySection, PolicySections } from './types';

export const policySection: z.ZodSchema<PolicySection> = z
  .object({
    id: z.string(),
    title: z.string(),
    children: z.lazy(() => z.array(policySection)).optional(),
    htmlContent: z.union([z.string(), z.array(z.string())]).optional(),
    miscData: z.record(z.unknown()).optional(),
  })
  .nonstrict();

export const policySections: z.ZodSchema<PolicySections> =
  z.array(policySection);
