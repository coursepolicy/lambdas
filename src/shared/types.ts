import middy from '@middy/core';
import { APIGatewayEvent } from 'aws-lambda';
import { SurveyResponsesRequestBody } from '../http/responses/schema';

// omit body

export type ExtendedApiGateWayEvent = Omit<
  APIGatewayEvent,
  'queryStringParameters'
> & {
  queryStringParameters: SurveyResponsesRequestBody;
};

export interface MiddlewareRequest
  extends middy.Request<ExtendedApiGateWayEvent> {}

export interface ResponseObject {
  labels: Labels;
  values: Values;
  responseId: string;
  displayedFields: string[];
  displayedValues: DisplayedValues;
}

export interface Labels {
  [index: string]: any;
  QID8: string;
  QID15: string;
  QID22: string;
  QID24: string;
  QID25: string[];
  QID26: string;
  QID28: string;
  status: string;
  QID16_1?: string;
  QID16_2?: string;
  QID16_3?: string;
  QID16_4?: string;
  QID16_5?: string;
  QID16_6?: string;
  QID16_7?: string;
  QID16_8?: string;
  QID16_9?: string;
  QID4_DO: string[];
  QID8_DO: string[];
  QID12_DO: string[];
  QID15_DO: string[];
  QID16_10?: string;
  QID16_DO?: string[];
  QID22_DO: string[];
  QID24_DO: string[];
  QID25_DO: string[];
  QID26_DO: string[];
  QID28_DO: string[];
  finished: string;
}

interface Values {
  [index: string]: any;
  QID8: number;
  QID15: number;
  QID22: number;
  QID24: number;
  QID25: string[];
  QID26: number;
  QID28: number;
  QID4_1: string;
  QID4_2: string;
  QID4_3: string;
  QID4_4: string;
  status: number;
  QID12_1: string;
  QID12_2: string;
  QID12_3: string;
  QID12_4: string;
  QID16_1?: number;
  QID16_2?: number;
  QID16_3?: number;
  QID16_4?: number;
  QID16_5?: number;
  QID16_6?: number;
  QID16_7?: number;
  QID16_8?: number;
  QID16_9?: number;
  QID4_DO: string[];
  QID8_DO: string[];
  endDate: string;
  QID12_DO: string[];
  QID15_DO: string[];
  QID16_10?: number;
  QID16_DO?: string[];
  QID22_DO: string[];
  QID24_DO: string[];
  QID25_DO: string[];
  QID26_DO: string[];
  QID28_DO: string[];
  duration: number;
  finished: number;
  progress: number;
  QID3_TEXT: string;
  ipAddress: string;
  startDate: string;
  QID13_TEXT: string;
  QID17_TEXT: string;
  QID19_TEXT?: string;
  QID20_TEXT?: string;
  QID25_6_TEXT?: string;
  QID30_TEXT: string;
  recordedDate: string;
  userLanguage: string;
  Q_RecaptchaScore: number;
  locationLatitude: string;
  locationLongitude: string;
  distributionChannel: string;
}

interface DisplayedValues {
  QID8: number[];
  QID15: number[];
  QID22: number[];
  QID24: number[];
  QID25: string[];
  QID26: number[];
  QID28: number[];
  QID16_1?: number[];
  QID16_2?: number[];
  QID16_3?: number[];
  QID16_4?: number[];
  QID16_5?: number[];
  QID16_6?: number[];
  QID16_7?: number[];
  QID16_8?: number[];
  QID16_9?: number[];
  QID16_10?: number[];
}

export interface GenerativeAiPolicy {
  courseNumber: string; // q4_3
  courseTitle: string; // q4_4
  instructor: string; // q4_1
  email: string; // q4_2
  courseDescription: string; // q3 text
  overallPolicy: string; // values.QID15,
  overallPolicyText: string; // none
  additionalPolicyText: string; // q12
  useCases?: UseCases; // q16
  ethicalGuidelines?: string[]; // q25
  additionalGuidelines?: string; // q26
  specificPoliciesForAssignments?: string; // q17 // specific generative ai policies
  additionalGenerativeAiToolsDeclarations?: string[]; // q26
  generativeAiToolDeclarations?: string[]; // q26
  campusWidePolicy?: string;
  departmentWidePolicy?: string;
  academicIntegrityPolicy?: string;
  otherPolicies?: string;
  additionalNotes: string;
  generatedAt: string;
}

export interface UseCases {
  reasonable: UseCaseEntry[];
  unreasonable: UseCaseEntry[];
}

interface UseCaseEntry {
  label: string;
  text: string;
}

interface PolicyMapper {
  [index: number]: string;
}

export const generativeAiPolicyMapper: PolicyMapper = {
  1: 'No restrictions',
  2: 'Allowed under conditions',
  3: 'Strictly prohibited',
} as const;

export interface Section {
  [key: string]: any;
  id: string;
  sectionTitle: string;
  subSections: SubSection[];
}

export interface SubSection {
  id: string;
  subSectionTitle: string;
  content: string;
  miscData?: Record<string, any>;
}

export type CourseAiPolicy = Section[];

export type CourseAiPolicyResponse = {
  header: string;
  content: CourseAiPolicy;
};
