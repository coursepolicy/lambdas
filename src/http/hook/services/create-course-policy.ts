import { v4 as uuidv4 } from 'uuid';
import {
  CourseAiPolicy,
  CourseAiPolicyResponse,
  GenerativeAiPolicy,
  SubSection,
} from '../../../shared';
import { format } from 'date-fns';

const policyFormatter = (
  courseAiPolicy: CourseAiPolicy,
  {
    courseTitle,
    courseNumber,
    generatedAt,
    email,
    instructor,
  }: GenerativeAiPolicy
): CourseAiPolicyResponse => {
  return {
    header: `
      <div>
        <h1>
          ${courseNumber}: ${courseTitle}
        </h1>
        <p>
          Course Instructor: ${instructor}[${email}]
          <span>Generated on ${format(new Date(generatedAt), 'PPP')}</span>
        </p>
      </div>
    `,
    content: courseAiPolicy,
  };
};

export const createCoursePolicy = (
  response: GenerativeAiPolicy
): CourseAiPolicyResponse => {
  const courseDescriptionSubSections: SubSection[] = [];
  const generativeAiPolicySubSections: SubSection[] = [];
  const additionalPoliciesSubSections: SubSection[] = [];

  // courseDescriptionSubSections
  courseDescriptionSubSections.push({
    id: uuidv4(),
    subSectionTitle: 'Introduction',
    content: `
    <section class="course-description-introduction-section">
      <h3>Course Description</h3>
      <p>${response.courseDescription}</p>
    </section>
    `,
  });
  // end of courseDescriptionSubSections

  // generativeAiPolicySubSections

  generativeAiPolicySubSections.push({
    id: uuidv4(),
    subSectionTitle: 'Introduction',
    miscData: { overallPolicy: response.overallPolicy },
    content: `
      <section class="policy-introduction-section">
        <h2>1. ${response.courseNumber} Generative AI Policy</h2>
        <p>${response.overallPolicyText}</p>
      </section>
    `,
  });

  if (response.useCases) {
    generativeAiPolicySubSections.push({
      id: uuidv4(),
      subSectionTitle: 'Use Cases',
      content: [
        `<div>
          <h3>Reasonable Use Cases ✅</h3>
          ${
            response.useCases.reasonable.length
              ? `
            ${response.useCases.reasonable.reduce(
              (acc, entry) =>
                (acc += `
                <strong>${entry.label}</strong>
                <ul>
                  <li>
                    
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
              ''
            )}
          `
              : '<p>None</p>'
          }
        </div>`,
        `<div>
          <h3>Unreasonable Use Cases ❌</h3>
          ${
            response.useCases.unreasonable.length
              ? `
            ${response.useCases.unreasonable.reduce(
              (acc, entry) =>
                (acc += `
                <strong>${entry.label}</strong>
                <ul>
                  <li>
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
              ''
            )}
          `
              : '<p>None</p>'
          }
        </div>`,
      ],
    });
  }

  if (response.specificPoliciesForAssignments) {
    generativeAiPolicySubSections.push({
      id: uuidv4(),
      subSectionTitle: 'Assignment Specific AI Policies',
      content: `
        <section class="policy-assignment-specific-section">
          <h3>Assignment/Project Specific AI Policies</h3>
          <p>${response.specificPoliciesForAssignments}</p>
        </section>
      `,
    });
  }

  if (response.ethicalGuidelines) {
    generativeAiPolicySubSections.push({
      id: uuidv4(),
      subSectionTitle: 'Ethical Guidelines',
      content: `
      <section class="policy-ehtical-guidelines-section">
        <h3>Ethical guidelines for using generative AI for this course:</h3>
        <ul>
          ${response.ethicalGuidelines.reduce(
            (acc, text) => (acc += `<li>${text}</li>`),
            ''
          )}
          ${
            response.additionalGuidelines
              ? `<li>${response.additionalGuidelines}</li>`
              : ''
          }
        </ul>
      </section>
      `,
    });
  }

  if (response.generativeAiToolDeclarations) {
    generativeAiPolicySubSections.push({
      id: uuidv4(),
      subSectionTitle: 'Declaration',
      content: `
        <section class="policy-declaration-section">
          <h3>How to declare the use of generative tools:</h3>
          <ul>
            ${response.generativeAiToolDeclarations.reduce(
              (acc, text) => (acc += `<li>${text}</li>`),
              ''
            )}
            ${
              response.additionalGenerativeAiToolsDeclarations
                ? `<li>${response.additionalGenerativeAiToolsDeclarations}</li>`
                : ''
            }
          </ul>
        </section>
      `,
    });
  }

  generativeAiPolicySubSections.push({
    id: uuidv4(),
    subSectionTitle: 'Additional Notes',
    content: `
     <section class="policy-notes-section">
        <h3>Additional Notes</h3>
        ${
          response.additionalNotes
            ? `
        <ul>
          <li>${response.additionalNotes}</li>
        </ul>
        `
            : ''
        }
      </section>
    `,
  });

  // end of generativeAiPolicySubSections

  // additionalPoliciesSubSections
  additionalPoliciesSubSections.push({
    id: uuidv4(),
    subSectionTitle: 'Introduction',
    content: `
      <section class="additional-policies-introduction-section">
        <h2>2. Additional Policies</h2>
        <p>${response.additionalPolicyText}</p>
      </section>
    `,
  });

  if (response.overallPolicy !== 'No restrictions') {
    additionalPoliciesSubSections.push({
      id: uuidv4(),
      subSectionTitle: 'Policy Links',
      content: `
        <section class="additional-policies-policy-links-section">
            <ul>
              ${
                response.campusWidePolicy
                  ? `
                
                <li>
                ${
                  response.campusWidePolicy.length
                    ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.campusWidePolicy}">Campus-wide generative AI policy</a></p>`
                    : 'Campus-wide generative AI policy: N/A'
                }
                </li>
              `
                  : ''
              }
              ${
                response.departmentWidePolicy
                  ? `
                <li>
                ${
                  response.departmentWidePolicy.length
                    ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.departmentWidePolicy}">Department-wide generative AI policy</a></p>`
                    : 'Department-wide generative AI policy: N/A'
                }
                </li>
              `
                  : ''
              }
              ${
                response.academicIntegrityPolicy
                  ? `
                  
                  <li>
                  ${
                    response.academicIntegrityPolicy.length
                      ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.academicIntegrityPolicy}">Academic Integrity policy</a></p>`
                      : 'Academic Integrity policy: N/A'
                  }
                  </li>
              `
                  : ''
              }
              ${
                response.otherPolicies
                  ? `
                  <li>
                  ${
                    response.otherPolicies.length
                      ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.otherPolicies}">Other policies</a></p>`
                      : 'Other policies: N/A'
                  }
                  </li>
              `
                  : ''
              }
            </ul>
          </section>
      
      `,
    });
  }
  // end of additionalPoliciesSubSections

  return policyFormatter(
    [
      {
        // section
        id: uuidv4(),
        sectionTitle: 'Course Description',
        subSections: courseDescriptionSubSections,
      },
      {
        id: uuidv4(),
        sectionTitle: 'Generative AI Policy',
        subSections: generativeAiPolicySubSections,
      },
      {
        id: uuidv4(),
        sectionTitle: 'Additional Policies',
        subSections: additionalPoliciesSubSections,
      },
    ],
    response
  );
};
