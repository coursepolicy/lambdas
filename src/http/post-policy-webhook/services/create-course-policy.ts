import { ulid } from 'ulid';
import {
  PolicySections,
  AiPolicy,
  MappedSurveyResponse,
  PolicySection,
  UseCases,
  HgseUseCases,
} from '../../../shared';

const policyFormatter = <T extends UseCases | HgseUseCases>(
  courseAiPolicy: PolicySections,
  { courseTitle, courseNumber, email, instructor }: MappedSurveyResponse<T>,
  organization?: string
): AiPolicy => {
  return {
    heading: `
      <div>
        <h1>
          AI Policy for ${courseNumber}: ${courseTitle}
        </h1>
        <p>
          Course Instructor: ${instructor} <a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="mailto:${email}">${email}</a>
        </p>
      </div>
    `
      .replace(/\n/g, '')
      .replace(/>(\s+)</g, '><')
      .trim(),
    sections: courseAiPolicy,
    metadata: {
      organization: organization || null
    }
  };
};

export const createCoursePolicy = <T extends UseCases | HgseUseCases>(organization?: string) => (
  response: MappedSurveyResponse<T>,
  test = false
): AiPolicy => {
  const courseDescriptionSubSections: PolicySection[] = [];
  const generativeAiPolicySubSections: PolicySection[] = [];
  const additionalPoliciesSubSections: PolicySection[] = [];

  // courseDescriptionSubSections
  courseDescriptionSubSections.push({
    id: test ? 'mockId' : ulid(),
    title: 'Introduction',
    htmlContent: `
        <h3>Course Description</h3>
        <p>${response.courseDescription}</p>
      `
      .replace(/\n/g, '')
      .replace(/>(\s+)</g, '><')
      .trim(),
  });
  // end of courseDescriptionSubSections

  // generativeAiPolicySubSections

  generativeAiPolicySubSections.push({
    id: test ? 'mockId' : ulid(),
    title: 'Introduction',
    miscData: { overallPolicy: response.overallPolicy },
    htmlContent: `<h2>1. ${
      response.courseNumber
    } Generative AI Policy</h2><p>${response.overallPolicyText.trim()}</p>`
      .replace(/\n/g, '')
      .replace(/>(\s+)</g, '><')
      .trim(),
  });

  if (response.useCases) {
    if ('reasonable' in response.useCases) {
      generativeAiPolicySubSections.push({
        id: test ? 'mockId' : ulid(),
        title: 'Use Cases',
        htmlContent: [
          `<h3>Reasonable Use Cases ✅</h3>
          ${
            response.useCases.reasonable.length
              ? `
            ${response.useCases.reasonable.reduce(
                (acc, entry) =>
                  (acc += `
                <p><strong>${entry.label}</strong></p>
                <ul>
                  <li>
                    
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
                ''
              )}
          `
              : '<p>None</p>'
          }`
            .replace(/\n/g, '')
            .replace(/>(\s+)</g, '><')
            .trim(),
          `
          <h3>Unreasonable Use Cases ❌</h3>
          ${
            response.useCases.unreasonable.length
              ? `
            ${response.useCases.unreasonable.reduce(
                (acc, entry) =>
                  (acc += `
                <p><strong>${entry.label}</strong></p>
                <ul>
                  <li>
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
                ''
              )}
          `
              : '<p>None</p>'
          }`
            .replace(/\n/g, '')
            .replace(/>(\s+)</g, '><')
            .trim(),
        ],
      });

    } else if ('acceptable' in response.useCases) {
      generativeAiPolicySubSections.push({
        id: test ? 'mockId' : ulid(),
        title: 'Use Cases',
        htmlContent: [
          `<h3>Acceptable Use Cases ✅</h3>
          ${
            response.useCases.acceptable.length
              ? `
            ${response.useCases.acceptable.reduce(
                (acc, entry) =>
                  (acc += `
                <p><strong>${entry.label}</strong></p>
                <ul>
                  <li>  
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
                ''
              )}
          `
              : '<p>None</p>'
          }`
            .replace(/\n/g, '')
            .replace(/>(\s+)</g, '><')
            .trim(),
          `
          <h3>Unacceptable Use Cases ❌</h3>
          ${
            response.useCases.unacceptable.length
              ? `
            ${response.useCases.unacceptable.reduce(
                (acc, entry) =>
                  (acc += `
                <p><strong>${entry.label}</strong></p>
                <ul>
                  <li>
                    <p>${entry.text}</p>
                  </li>
                </ul>
              `),
                ''
              )}
          `
              : '<p>None</p>'
          }`
            .replace(/\n/g, '')
            .replace(/>(\s+)</g, '><')
            .trim(),
        ],
      });

    }
  }

  if (response.specificPoliciesForAssignments) {
    generativeAiPolicySubSections.push({
      id: test ? 'mockId' : ulid(),
      title: 'Assignment Specific AI Policies',
      htmlContent:
        `<h3>Assignment/Project Specific AI Policies</h3><p>${response.specificPoliciesForAssignments}</p>`
          .replace(/\n/g, '')
          .replace(/>(\s+)</g, '><')
          .trim(),
    });
  }

  if (response.ethicalGuidelines) {
    generativeAiPolicySubSections.push({
      id: test ? 'mockId' : ulid(),
      title: 'Ethical Guidelines',
      htmlContent:
        `<h3>Ethical guidelines for using generative AI for this course:</h3>
        <ul>
          ${response.ethicalGuidelines.reduce(
            (acc, text) => (acc += `<li>${text}</li>`),
            ''
          )}
          ${
            response.additionalGuidelines
              ? `<li>${response.additionalGuidelines}</li>`
              : ''
          }
        </ul>
      `
          .replace(/\n/g, '')
          .replace(/>(\s+)</g, '><')
          .trim(),
    });
  }

  if (response.generativeAiToolDeclarations) {
    generativeAiPolicySubSections.push({
      id: test ? 'mockId' : ulid(),
      title: 'Declaration',
      htmlContent: `<h3>How to declare the use of generative tools:</h3>
          <ul>
            ${response.generativeAiToolDeclarations.reduce(
              (acc, text) => (acc += `<li>${text}</li>`),
              ''
            )}
            ${
              response.additionalGenerativeAiToolsDeclarations
                ? `<li>${response.additionalGenerativeAiToolsDeclarations}</li>`
                : ''
            }
          </ul>`
        .replace(/\n/g, '')
        .replace(/>(\s+)</g, '><')
        .trim(),
    });
  }

  generativeAiPolicySubSections.push({
    id: test ? 'mockId' : ulid(),
    title: 'Additional Notes',
    htmlContent: `
        <h3>Additional Notes</h3>
        ${
          response.additionalNotes
            ? `
        <ul>
          <li>${response.additionalNotes}</li>
        </ul>
        `
            : ''
        }`
      .replace(/\n/g, '')
      .replace(/>(\s+)</g, '><')
      .trim(),
  });

  // end of generativeAiPolicySubSections

  // additionalPoliciesSubSections
  additionalPoliciesSubSections.push({
    id: test ? 'mockId' : ulid(),
    title: 'Introduction',
    htmlContent:
      `<h2>2. Additional Policies</h2><p>${response.additionalPolicyText}</p>`
        .replace(/\n/g, '')
        .replace(/>(\s+)</g, '><')
        .trim(),
  });

  // check to see if all the policy links exists and are not empty
  if (response.overallPolicy !== 'No restrictions') {
    additionalPoliciesSubSections.push({
      id: test ? 'mockId' : ulid(),
      title: 'Policy Links',
      htmlContent: `<ul>
              ${
                response.campusWidePolicy
                  ? `
                <li>
                ${
                  response.campusWidePolicy.length
                    ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.campusWidePolicy}">Campus-wide generative AI policy</a></p>`
                    : 'Campus-wide generative AI policy: N/A'
                }
                </li>
              `
                  : ''
              }
              ${
                response.departmentWidePolicy
                  ? `
                <li>
                ${
                  response.departmentWidePolicy.length
                    ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.departmentWidePolicy}">Department-wide generative AI policy</a></p>`
                    : 'Department-wide generative AI policy: N/A'
                }
                </li>
              `
                  : ''
              }
              ${
                response.academicIntegrityPolicy
                  ? `
                  <li>
                  ${
                    response.academicIntegrityPolicy.length
                      ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.academicIntegrityPolicy}">Academic Integrity policy</a></p>`
                      : 'Academic Integrity policy: N/A'
                  }
                  </li>
              `
                  : ''
              }
              ${
                response.otherPolicies
                  ? `
                  <li>
                  ${
                    response.otherPolicies.length
                      ? `<p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="${response.otherPolicies}">Other policies</a></p>`
                      : 'Other policies: N/A'
                  }
                  </li>
              `
                  : ''
              }
            </ul>`
        .replace(/\n/g, '')
        .replace(/>(\s+)</g, '><')
        .trim(),
    });
  }
  // end of additionalPoliciesSubSections

  return policyFormatter(
    [
      {
        // section
        id: test ? 'mockId' : ulid(),
        title: 'Course Description',
        children: courseDescriptionSubSections,
      },
      {
        id: test ? 'mockId' : ulid(),
        title: 'Generative AI Policy',
        children: generativeAiPolicySubSections,
      },
      {
        id: test ? 'mockId' : ulid(),
        title: 'Additional Policies',
        children: additionalPoliciesSubSections,
      },
    ],
    response,
    organization
  );
};
