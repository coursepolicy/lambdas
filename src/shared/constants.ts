export const qualtricsBaseUrl = 'https://sjc1.qualtrics.com/API/v3';

export const useCaseReasonabilityMapper: {
  [index: number]: string;
} = {
  1: 'Reasonable',
  2: 'Not Reasonable',
  3: 'Not Applicable',
} as const;

export const useCasesMapper: {
  [index: string]: string;
} = {
  'Grammar Check': 'QID16_1',
  'Concept Learning': 'QID16_2',
  'Argument Generation': 'QID16_3',
  'Outline Generation': 'QID16_12',
  'Essay Generation': 'QID16_4',
  'Test-taking': 'QID16_5',
  'Literature Discovery': 'QID16_6',
  'Summary Generation': 'QID16_7',
  'Coding Assistant': 'QID16_8',
  'Art Creation': 'QID16_9',
  'Data Fabrication': 'QID16_10',
} as const;

export const generativeAiPolicyMapper: {
  [index: number]: string;
} = {
  1: 'No restrictions',
  2: 'Allowed under conditions',
  3: 'Strictly prohibited',
} as const;
