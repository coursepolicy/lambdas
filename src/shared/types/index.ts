import middy from "@middy/core";
import { APIGatewayEvent } from "aws-lambda";
import { SurveyHookRequestBody } from "../../hook/schema";
import { SurveyResponsesRequestBody } from "../../responses/schema";

// omit body

export type ExtendedApiGateWayEvent = APIGatewayEvent & {
  parsedBody: SurveyHookRequestBody;
  parsedQueryParams: SurveyResponsesRequestBody;
};

export interface MiddlewareRequest
  extends middy.Request<ExtendedApiGateWayEvent> {}
