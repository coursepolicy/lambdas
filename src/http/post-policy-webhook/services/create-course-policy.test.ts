import { MappedSurveyResponse } from '../../../shared';
import { createCoursePolicy } from './create-course-policy';

describe('createCoursePolicy', () => {
  it('should return consistent output', () => {
    const mockResponse: MappedSurveyResponse = {
      id: 'mockId',
      createdAt: '2021-01-01T00:00:00.000Z',
      courseTitle: 'Test Course',
      courseNumber: '101',
      email: 'test@email.com',
      instructor: 'Test Instructor',
      courseDescription: 'This is a test course.',
      overallPolicy: 'Some Policy',
      overallPolicyText: 'Some policy text.',

      useCases: {
        reasonable: [
          {
            label: 'Test Reasonable Use Case',
            text: 'Description for a reasonable use case.',
          },
        ],
        unreasonable: [
          {
            label: 'Test Unreasonable Use Case',
            text: 'Description for an unreasonable use case.',
          },
        ],
      },

      specificPoliciesForAssignments:
        'Specific policies related to assignments for the test course.',

      ethicalGuidelines: [
        'First ethical guideline.',
        'Second ethical guideline.',
      ],

      additionalGuidelines:
        'Additional ethical guidelines for the test course.',

      generativeAiToolDeclarations: [
        'First declaration regarding generative tools.',
        'Second declaration regarding generative tools.',
      ],

      additionalGenerativeAiToolsDeclarations: [
        'Additional declarations regarding generative AI tools for the test course.',
      ],

      additionalNotes: 'Some additional notes regarding the test course.',

      additionalPolicyText: 'Additional policies text for the test course.',

      campusWidePolicy: 'https://example.com/campus-wide-policy',
      departmentWidePolicy: 'https://example.com/department-wide-policy',
      academicIntegrityPolicy: 'https://example.com/academic-integrity-policy',
      otherPolicies: 'https://example.com/other-policies',
    };

    const result = createCoursePolicy(mockResponse, true);
    expect(result).toMatchSnapshot();
  });
});
