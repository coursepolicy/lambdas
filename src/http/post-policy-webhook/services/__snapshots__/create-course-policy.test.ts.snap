// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HGSE createCoursePolicy should return consistent HGSE output 1`] = `
{
  "heading": "<div><h1>          AI Policy for 101: Test Course        </h1><p>          Course Instructor: Test Instructor <a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="mailto:test@email.com">test@email.com</a></p></div>",
  "metadata": {
    "organization": "harvard",
  },
  "sections": [
    {
      "children": [
        {
          "htmlContent": "<h3>Course Description</h3><p>This is a test course.</p>",
          "id": "mockId",
          "title": "Introduction",
        },
      ],
      "id": "mockId",
      "title": "Course Description",
    },
    {
      "children": [
        {
          "htmlContent": "<h2>1. 101 Generative AI Policy</h2><p>Some policy text.</p>",
          "id": "mockId",
          "miscData": {
            "overallPolicy": "Some Policy",
          },
          "title": "Introduction",
        },
        {
          "htmlContent": [
            "<h3>Acceptable Use Cases ✅</h3><p><strong>Test Acceptable Use Case</strong></p><ul><li><p>Description for a acceptable use case.</p></li></ul>",
            "<h3>Unacceptable Use Cases ❌</h3><p><strong>Test Unacceptable Use Case</strong></p><ul><li><p>Description for an unacceptable use case.</p></li></ul>",
          ],
          "id": "mockId",
          "title": "Use Cases",
        },
        {
          "htmlContent": "<h3>Assignment/Project Specific AI Policies</h3><p>Specific policies related to assignments for the test course.</p>",
          "id": "mockId",
          "title": "Assignment Specific AI Policies",
        },
        {
          "htmlContent": "<h3>Ethical guidelines for using generative AI for this course:</h3><ul><li>First ethical guideline.</li><li>Second ethical guideline.</li><li>Additional ethical guidelines for the test course.</li></ul>",
          "id": "mockId",
          "title": "Ethical Guidelines",
        },
        {
          "htmlContent": "<h3>How to declare the use of generative tools:</h3><ul><li>First declaration regarding generative tools.</li><li>Second declaration regarding generative tools.</li><li>Additional declarations regarding generative AI tools for the test course.</li></ul>",
          "id": "mockId",
          "title": "Declaration",
        },
        {
          "htmlContent": "<h3>Additional Notes</h3><ul><li>Some additional notes regarding the test course.</li></ul>",
          "id": "mockId",
          "title": "Additional Notes",
        },
      ],
      "id": "mockId",
      "title": "Generative AI Policy",
    },
    {
      "children": [
        {
          "htmlContent": "<h2>2. Additional Policies</h2><p>Additional policies text for the test course.</p>",
          "id": "mockId",
          "title": "Introduction",
        },
        {
          "htmlContent": "<ul><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/campus-wide-policy">Campus-wide generative AI policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/department-wide-policy">Department-wide generative AI policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/academic-integrity-policy">Academic Integrity policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/other-policies">Other policies</a></p></li></ul>",
          "id": "mockId",
          "title": "Policy Links",
        },
      ],
      "id": "mockId",
      "title": "Additional Policies",
    },
  ],
}
`;

exports[`createCoursePolicy should return consistent output 1`] = `
{
  "heading": "<div><h1>          AI Policy for 101: Test Course        </h1><p>          Course Instructor: Test Instructor <a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="mailto:test@email.com">test@email.com</a></p></div>",
  "metadata": {
    "organization": null,
  },
  "sections": [
    {
      "children": [
        {
          "htmlContent": "<h3>Course Description</h3><p>This is a test course.</p>",
          "id": "mockId",
          "title": "Introduction",
        },
      ],
      "id": "mockId",
      "title": "Course Description",
    },
    {
      "children": [
        {
          "htmlContent": "<h2>1. 101 Generative AI Policy</h2><p>Some policy text.</p>",
          "id": "mockId",
          "miscData": {
            "overallPolicy": "Some Policy",
          },
          "title": "Introduction",
        },
        {
          "htmlContent": [
            "<h3>Reasonable Use Cases ✅</h3><p><strong>Test Reasonable Use Case</strong></p><ul><li><p>Description for a reasonable use case.</p></li></ul>",
            "<h3>Unreasonable Use Cases ❌</h3><p><strong>Test Unreasonable Use Case</strong></p><ul><li><p>Description for an unreasonable use case.</p></li></ul>",
          ],
          "id": "mockId",
          "title": "Use Cases",
        },
        {
          "htmlContent": "<h3>Assignment/Project Specific AI Policies</h3><p>Specific policies related to assignments for the test course.</p>",
          "id": "mockId",
          "title": "Assignment Specific AI Policies",
        },
        {
          "htmlContent": "<h3>Ethical guidelines for using generative AI for this course:</h3><ul><li>First ethical guideline.</li><li>Second ethical guideline.</li><li>Additional ethical guidelines for the test course.</li></ul>",
          "id": "mockId",
          "title": "Ethical Guidelines",
        },
        {
          "htmlContent": "<h3>How to declare the use of generative tools:</h3><ul><li>First declaration regarding generative tools.</li><li>Second declaration regarding generative tools.</li><li>Additional declarations regarding generative AI tools for the test course.</li></ul>",
          "id": "mockId",
          "title": "Declaration",
        },
        {
          "htmlContent": "<h3>Additional Notes</h3><ul><li>Some additional notes regarding the test course.</li></ul>",
          "id": "mockId",
          "title": "Additional Notes",
        },
      ],
      "id": "mockId",
      "title": "Generative AI Policy",
    },
    {
      "children": [
        {
          "htmlContent": "<h2>2. Additional Policies</h2><p>Additional policies text for the test course.</p>",
          "id": "mockId",
          "title": "Introduction",
        },
        {
          "htmlContent": "<ul><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/campus-wide-policy">Campus-wide generative AI policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/department-wide-policy">Department-wide generative AI policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/academic-integrity-policy">Academic Integrity policy</a></p></li><li><p><a target="_blank" rel="noopener noreferrer nofollow" class="editor-links" href="https://example.com/other-policies">Other policies</a></p></li></ul>",
          "id": "mockId",
          "title": "Policy Links",
        },
      ],
      "id": "mockId",
      "title": "Additional Policies",
    },
  ],
}
`;
